### DISCUSSION TITLE
Update-only RLS with upsert()

### DISCUSSION THREAD

**Discussion by leogreu** [supabase/supabase#30499] [Questions]
Hello folks,

I have a problem that I just cannot solve:

I am using `upsert()` as [idempotent statement](https://jonmeyers.io/blog/use-on-conflict-to-upsert-in-postgresql/) for all data modifications. It is great for simplicity and robustness since I don’t need to track which data is new and which was updated.

Now, I want to have a RLS policy that only grants `UPDATE` to a specific user group (lets say a reviewer may only verify data points but may not create new ones).

The problem is that Postgres always starts with the `INSERT` part of `upsert()` (since it stands for `INSERT … ON CONFLICT DO UPDATE`). And since there is no successful `INSERT` RLS, it fails – even when the data exists, i.e., there would be a conflict. If the data would not exist, an error would actually be desirable – but if it does, it should succeed.

However, it seems impossible to implement this relatively simple use case.

The only solution I see is to differentiate between `insert()` and `update()`, but that would result in more complexity, error possibilities, and it would not support to batch persist multiple data points if some are new and some were updated.

Am I missing something here or do you have any creative idea to solve this problem?

**Comment by GaryAustin1** [comment]
Upsert requires meeting all 3 policies and no way around that as works as you describe. 
You might be able to move your data thru an rpc call that sorts the type of user and takes different actions.
